examples/proxy/proxy.ts

```diff
@@ -0,0 +1,72 @@
+import { TaskExecutor } from "@golem-sdk/golem-js";
+
+/**
+ * An example demonstrating the use of a proxy server to send and receive http requests to the provider.
+ * After starting the proxy server, you can make any http request
+ * that will be processed by the http server on the provider's machine, eg. `curl http://localhost`
+ */
+(async function main() {
+  const executor = await TaskExecutor.create({
+    package: "golem/node:20-alpine",
+    capabilities: ["vpn"],
+    networkIp: "192.168.0.0/24",
+    skipProcessSignals: true,
+
+    // Restart the HTTP server up to 5 times
+    maxTaskRetries: 5,
+
+    // If you're using TaskExecutor, you want the "task" to last long in that case
+    taskTimeout: 60 * 60 * 1000,
+    activityExecuteTimeout: 60 * 60 * 1000,
+  });
+
+  try {
+    await executor.run(async (ctx) => {
+      const PORT_ON_PROVIDER = 80;
+      const PORT_ON_REQUESTOR = 8080;
+
+      // Install the server script
+      await ctx.uploadFile(`./proxy/server.js`, "/golem/work/server.js");
+
+      // Start the server process on the provider
+      const server = await ctx.spawn(`PORT=${PORT_ON_PROVIDER} node /golem/work/server.js`);
+
+      server.stdout.on("data", (data) => console.log("provider>", data));
+      server.stderr.on("data", (data) => console.error("provider>", data));
+
+      // Create a proxy instance
+      const proxy = ctx.createTcpProxy(PORT_ON_PROVIDER);
+      proxy.events.on("error", (error) => console.error("TcpProxy reported an error:", error));
+
+      // Start listening and expose the port on your requestor machine
+
+      await proxy.listen(PORT_ON_REQUESTOR);
+
+      console.log(`Server Proxy listen at http://localhost:${PORT_ON_REQUESTOR}`);
+
+      // Prepare and register shutdown handlers for graceful termination
+      let isClosing = false;
+      const stopTask = async () => {
+        if (isClosing) {
+          console.log("Already closing, ignoring subsequent shutdown request");
+          return;
+        }
+
+        isClosing = true;
+
+        console.log("Shutting down gracefully");
+        await proxy.close();
+        await server.waitForExit();
```

Comment:
by default, after 30 seconds it will throw a `TimeoutError` because the `server` (process) will not end. We cannot send any `SIGTERM` or `SIGINT` signal to the process (on the provider/batch in activity), so the only way is to try to end the `task/workerFunction` without waiting for exit from remoteProcess. Then `TaskExecutor` will treat the task as completed and will terminate the activity and agreement. It's not a pretty solution, but I think it's the only possible one at this moment..

Reply:
I see your point. I've changed this line to leave a comment explaining that it's safe to leave the function after `await proxy.close()` to not run into this issue.

I've checked ya-client spec and it appears that we really can't interrupt a running batch via the API. On one hand this could be a feature request to the core team, but on the other, I think that in this context `WorkContext.spawn` is very poorly named and super misleading as it does not have a lot in common with the reality which it intended to capture. We can't even run a second command concurrently after we use `spawn`.

Thus, I think the last change which we could introduce here is a rename of this method and deprecation of the old name so that it can be removed in v3 or later.

```
WorkContext.run
WorkContext.runAndStream -> replaces: WorkContext.spawn
```

This way the `on('close', () => {})` handlers etc can still be used intuitively, and we don't give any false hopes to `spawn` users.